File: maps-hasher.py
# SPDX-License-Identifier: WTFPL
# Copyright 2022-2024 rtldg <rtldg@protonmail.com>

import bz2
import csv
import glob
import hashlib
import mmap
import os
import shutil
from pathlib import Path
from stat import *
from datetime import datetime

def print_and_to_shit(s):
    print(s)
    with open("shit.txt", "a", encoding="utf-8") as shit:
        shit.write(s + "\n")

def normal_name(m):
    return m.strip().replace('.', '_').lower()

def main(csvname, automatic, mapsfolder, timestampFixer, skipExistingHash, canonClobberCheck):
    if not automatic and os.path.exists(csvname) and os.path.getsize(csvname) > 50:
        raise Exception("DONT OVERWRITE THAT CSV!")

    existing_canon = {}
    if canonClobberCheck:
        canonClobber = open("canonClobber.csv", "a", encoding="utf-8")
        with open("processed/main.fastdl.me/maps_index.html.csv", newline='', encoding="utf-8") as f:
            for line in csv.reader(f):
                existing_canon[normal_name(line[0])] = line[1]

    existing_names = {}
    existing_recents = {}
    if automatic:
        # this will fail if you haven't ran `python process.py` yet
        with open("processed/main.fastdl.me/hashed_index.html.csv", newline='', encoding="utf-8") as f:
            for line in csv.reader(f):
                existing_names[normal_name(line[0])] = line[4] # url, hopefully from gb
        # only allow clobbering existing map names for recently added gamebanana downloads...
        with open("recently_added.csv", newline='', encoding="utf-8") as f:
            for line in csv.reader(f):
                if line[0] == "mapname":
                    continue
                existing_recents[line[0]] = line[4].split("_")[0]
                if line[0] in existing_names:
                    # lolololololol... this entire site is such a hack...
                    if ("https://gamebanana.com/mods/"+line[4].split("_")[0]) == existing_names[line[0]]:
                        del existing_names[line[0]]

    newly_hashed = []
    with open(csvname, ("a" if automatic else "w"), newline="", encoding="utf-8") as csvfile:
        mycsv = csv.writer(csvfile)
        if not automatic:
            mycsv.writerow(["mapname","filesize","filesize_bz2","sha1","note"])
        for filename in glob.iglob(mapsfolder + "/**/*.bsp", recursive=True):
            statttt = os.stat(filename)
            if S_ISDIR(statttt.st_mode):
                continue
            filesize = statttt.st_size
            if filesize == 0:
                print_and_to_shit(f"==== empty file {filename}")
                continue
            with open(filename, "rb") as f:
                mm = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
                vbspver = mm.read(5)
                if vbspver != b'VBSP\x13' and vbspver != b'VBSP\x14':
                    thing = ""
                    if vbspver == b'VBSP\x15':
                        thing = "==== skipping CS:GO map " + filename
                    elif vbspver == b'VBSP\x19':
                        thing = "==== skipping Momentum Mod / Strata Source map " + filename
                    elif vbspver[:4] == b'\x1E\x00\x00\x00' or vbspver[:4] == b'\x1D\x00\x00\x00':
                        thing = "==== skipping GoldSrc map? " + filename
                    else:
                        thing = "==== not a CS:S map? " + filename
                    print_and_to_shit(thing)
                    continue
                mm.seek(0)
                digest = hashlib.sha1(mm).hexdigest()
                #print("Hash {} -- {}".format(digest, filename))
                renameto = "../hashed/" + digest + ".bsp"
                exists = os.path.exists(renameto)
                if not exists:
                    """
                    if timestampFixer:
                        print("wtf bad??? {} {}".format(filename, renameto))
                        mm.close()
                        continue
                    """
                    print("copying new! {} -> {}".format(filename, renameto))
                    shutil.copy2(filename, renameto)
                    with bz2.open(renameto + ".bz2", "wb") as fbz2:
                        mm.seek(0)
                        unused = fbz2.write(mm)
                    shutil.copystat(renameto, renameto+".bz2")
                mm.close()
                if exists:
                    if os.stat(renameto).st_size != filesize:
                        print_and_to_shit(f"??????? HASH COLLISION WOW {digest} {filename}")
                        continue
                    mtimeThis = os.path.getmtime(filename)
                    mtimeHashed = os.path.getmtime(renameto)
                    if mtimeThis < mtimeHashed:
                        if timestampFixer:
                            print("timestamping {} from {} to {}".format(digest, datetime.utcfromtimestamp(mtimeHashed), datetime.utcfromtimestamp(mtimeThis)))
                            os.utime(renameto, (mtimeThis, mtimeThis))
                            os.utime(renameto+".bz2", (mtimeThis, mtimeThis))
                            continue
                        else:
                            print("older timestamp for {} from {} -- {} -> {}!".format(digest, filename, datetime.utcfromtimestamp(mtimeHashed), datetime.utcfromtimestamp(mtimeThis)))
                if exists and skipExistingHash:
                    continue
                filesize_bz2 = os.stat(renameto + ".bz2").st_size
                pp = Path(filename)
                stem = pp.stem
                if stem.startswith("#") and len(stem) > 1:
                    stem = stem[1:]
                row = [stem,filesize,filesize_bz2,digest,str(pp.parent).replace("\\", "/").replace(mapsfolder+"/", "")]
                if automatic and normal_name(row[0]) in existing_names:
                    if existing_recents.get(normal_name(row[0]), "mrbeast") != row[4].split("_")[0]:
                        row[0] = "#" + row[0]
                if canonClobberCheck and not automatic:
                    if row[0] in existing_canon and existing_canon[row[0]] != row[3]:
                        canonClobber.write(f"{row[0]},{existing_canon[row[0]]},\n")
                        print_and_to_shit(f"  ^^^^ name collision {digest} {filename} (existing: {row[0]} & {existing_canon[row[0]]}")
                if not exists:
                    newly_hashed.append(row)
                mycsv.writerow(row)
    if canonClobber != None:
        canonClobber.close()
    return newly_hashed

if __name__ == "__main__":
    main("unprocessed/misc3.csv", False, "../todo-gb/shit", False, False, True)
    #main("unprocessed/unloze-css_ze-unique.csv", False, "C:/shared/unloze/css_ze", False, True)
    #main("unprocessed/unloze-css_ze-all.csv", False, "C:/shared/unloze/css_ze", True, False)
    #main("unprocessed/moxx-terabox-unique.csv", False, "F:/terabox/cstrike_all/a", False, True)
    #main("unprocessed/moxx-terabox-all.csv", False, "F:/terabox/cstrike_all/a", True, False)
    #main("unprocessed/moxx-terabox-archives.csv", False, "F:/terabox/cstrike_all/archives", True, False)

File: mainframe/src/base/mapshasher.rs
// SPDX-License-Identifier: WTFPL
// Copyright 2025 rtldg <rtldg@protonmail.com>

#[cfg(windows)]
use std::os::windows::fs::FileTimesExt;

use std::{collections::HashMap, io::Write, path::Path, sync::Arc};

use crate::{
	SETTINGS,
	csv::{ProcessedCsvRow, UnprocessedCsvRow},
	gamebanana::GamebananaID,
	normalize_mapname,
};

use sha1::Digest;
use tokio::io::AsyncWriteExt;

#[derive(Debug, PartialEq, Eq)]
pub enum Mode {
	Automatic,
	Manual,
}

pub(crate) async fn run<P1: AsRef<Path>, P2: AsRef<Path>>(
	outcsvpath: P1,
	mode: Mode,
	mapsfolder: P2,
	timestamp_fixer: bool,
	skip_existing_hash: bool,
	canon_clobber_check: bool,
) -> anyhow::Result<Vec<UnprocessedCsvRow>> {
	run_inner(
		outcsvpath.as_ref(),
		mode,
		mapsfolder.as_ref(),
		timestamp_fixer,
		skip_existing_hash,
		canon_clobber_check,
	)
	.await
}

async fn run_inner(
	outcsvpath: &Path,
	mode: Mode,
	mapsfolder: &Path,
	timestamp_fixer: bool,
	skip_existing_hash: bool,
	canon_clobber_check: bool,
) -> anyhow::Result<Vec<UnprocessedCsvRow>> {
	let mapsfolder = mapsfolder.canonicalize()?;

	if mode == Mode::Manual {
		if let Ok(metadata) = tokio::fs::metadata(&outcsvpath).await {
			if metadata.len() > 75 {
				anyhow::bail!("DON'T OVERWRITE THAT CSV! ({})", outcsvpath.display());
			}
		}
	}

	let mut existing_canon = HashMap::new();
	if canon_clobber_check {
		// TODO (low) canonClobber.csv
		let mut maps_index_csv =
			csv::Reader::from_path(SETTINGS.dir_maps_cstrike.join("processed/main.fastdl.me/maps_index.html.csv"))?;
		let headers = maps_index_csv.headers()?.clone();
		let mut raw_record = csv::StringRecord::new();
		while maps_index_csv.read_record(&mut raw_record)? {
			let row = raw_record.deserialize::<ProcessedCsvRow>(Some(&headers))?;
			let _ = existing_canon.insert(normalize_mapname(row.mapname), row.sha1.to_string());
		}
	}

	let mut existing_names = HashMap::<String, Option<GamebananaID>>::new();
	let mut existing_recents = HashMap::<String, Option<GamebananaID>>::new();
	if mode == Mode::Automatic {
		let mut hashed_csv = csv::Reader::from_path(
			SETTINGS
				.dir_maps_cstrike
				.join("processed/main.fastdl.me/hashed_index.html.csv"),
		)?;
		let headers = hashed_csv.headers()?.clone();
		let mut raw_record = csv::StringRecord::new();
		while hashed_csv.read_record(&mut raw_record)? {
			let row = raw_record.deserialize::<ProcessedCsvRow>(Some(&headers))?;
			let _ = existing_names.insert(
				normalize_mapname(row.mapname),
				row.url
					.split("https://gamebanana.com/mods/")
					.nth(1)
					.map(|id| id.parse().unwrap()),
			);
		}

		let mut recently_added_csv = csv::Reader::from_path(SETTINGS.dir_maps_cstrike.join("recently_added.csv"))?;
		// only allow clobbering mapnames for recently added gamebanana downloads....
		for row in recently_added_csv.deserialize::<UnprocessedCsvRow>() {
			let row = row?;
			let id = if let Some(note) = row.note {
				note.split('_').next().and_then(|s| s.parse().ok())
			} else {
				None
			};
			let mapname = normalize_mapname(&row.mapname);
			if let Some(existing_names_id) = existing_names.get(&mapname) {
				if id == *existing_names_id {
					existing_names.remove(&mapname);
				}
			}
			existing_recents.insert(mapname, id);
		}
	}

	let mut newly_hashed = vec![];
	let mut outcsv = csv::Writer::from_writer(vec![]);

	for entry in glob::glob(&mapsfolder.join("**/*.bsp").to_string_lossy())? {
		let entry = entry?.canonicalize()?;

		let metadata = tokio::fs::metadata(&entry).await?;
		if metadata.is_dir() {
			continue;
		}
		let filesize = metadata.len();
		if filesize == 0 {
			println!("==== empty file {}", entry.display());
			continue;
		}
		// the smallest known bsp (from gb) is 12k bytes
		if filesize < 5_000 {
			println!("==== file too small ({filesize} bytes) {}", entry.display());
			continue;
		}

		let content = Arc::new(tokio::fs::read(&entry).await?);

		let vbspver = &content[0..5];

		if vbspver != b"VBSP\x13" && vbspver != b"VBSP\x14" {
			if vbspver == b"VBSP\x15" {
				println!("==== skipping CS:GO map {}", entry.display());
			} else if vbspver == b"VBSP\x19" {
				println!("==== skipping Momentum Mod / Strata Source map {}", entry.display());
			} else if &vbspver[0..4] == b"\x1E\x00\x00\x00" || &vbspver[0..4] == b"\x1D\x00\x00\x00" {
				println!("==== skipping GoldSrc Map? {}", entry.display());
			} else {
				println!("==== not a CS:S map? {}", entry.display());
			}
			continue;
		}

		// arguably unnecessary to use spawn_blocking here...
		let digest = tokio::task::spawn_blocking({
			let content = Arc::clone(&content);
			move || const_hex::encode(sha1::Sha1::digest(content.as_slice()))
		})
		.await
		.unwrap();

		let hashedbsppath = SETTINGS.dir_hashed.join(format!("{digest}.bsp"));
		let hashedbz2path = SETTINGS.dir_hashed.join(format!("{digest}.bsp.bz2"));
		let exists = tokio::fs::try_exists(&hashedbz2path).await?;
		let filesize_bz2;

		if !exists {
			/* TODO: timestampFixer
			if timestampFixer:
				print("wtf bad??? {} {}".format(filename, renameto))
				mm.close()
				continue
			*/
			println!("copying new! {} -> {}", entry.display(), hashedbsppath.display());

			#[cfg(target_os="windows")]
			fn fill_filetimes(metadata: &std::fs::Metadata) -> anyhow::Result<std::fs::FileTimes> {
				Ok(std::fs::FileTimes::new().set_modified(metadata.modified()?).set_created(metadata.created()?))
			}
			#[cfg(not(target_os="windows"))]
			fn fill_filetimes(metadata: &std::fs::Metadata) -> anyhow::Result<std::fs::FileTimes> {
				Ok(std::fs::FileTimes::new().set_modified(metadata.modified()?))
			}
			let filetimes = fill_filetimes(&metadata)?;

			let copy_task = tokio::task::spawn_blocking({
				let content = Arc::clone(&content);
				let hashedbsppath = hashedbsppath.clone();
				move || {
					let mut file = std::fs::File::create_new(hashedbsppath)?;
					file.write_all(&content)?;
					file.flush()?;
					file.set_times(filetimes)?;
					std::io::Result::Ok(())
				}
			});

			let bz2_task = tokio::task::spawn_blocking({
				let content = Arc::clone(&content);
				let hashedbz2path = hashedbz2path.clone();
				move || {
					let mut outbuf = std::io::BufWriter::new(std::fs::File::create_new(hashedbz2path)?);
					let mut compressor = bzip2::read::BzEncoder::new(content.as_slice(), bzip2::Compression::best());
					let filesize_bz2 = std::io::copy(&mut compressor, &mut outbuf)?;
					outbuf.flush()?;
					outbuf.into_inner()?.set_times(filetimes)?;
					anyhow::Result::<u64, anyhow::Error>::Ok(filesize_bz2)
				}
			});

			copy_task.await.unwrap()?;
			filesize_bz2 = bz2_task.await.unwrap()?;
		} else {
			// TODO: Read size from processed files instead... Kind of a hassle though.
			/*
			if tokio::fs::metadata(&hashedbsppath).await?.len() != filesize {
				println!("?????? HASH COLLISION WOW {} {}", digest, entry.display());
				continue;
			}
			*/

			let bz2_metadata = tokio::fs::metadata(&hashedbz2path).await?;
			filesize_bz2 = bz2_metadata.len();

			let bz2_modified = bz2_metadata.modified()?;
			let new_modified = metadata.modified()?;

			if new_modified < bz2_modified {
				if timestamp_fixer {
					println!("timestamping {} from {:#?} to {:#?}", digest, bz2_modified, new_modified);
					let times = std::fs::FileTimes::new().set_modified(new_modified);
					std::fs::OpenOptions::new()
						.write(true)
						.open(&hashedbsppath)?
						.set_times(times)?;
					std::fs::OpenOptions::new()
						.write(true)
						.open(&hashedbz2path)?
						.set_times(times)?;
					// TODO: Don't continue?
					continue;
				} else {
					// TODO: fix timestamp format
					println!(
						"older timestamp for {} from {} -- {:#?} -> {:#?}!",
						digest,
						entry.display(),
						bz2_modified,
						new_modified
					);
				}
			}

			if skip_existing_hash {
				continue;
			}
		}

		let mut stem = entry.file_stem().unwrap().to_str().unwrap().to_owned();
		if stem == "#" {
			// stupid
			stem = rand::random::<u64>().to_string();
		}
		if stem.starts_with('#') && stem.len() > 1 {
			stem.remove(0);
		}

		let parent = if mode == Mode::Manual {
			entry.parent().unwrap().strip_prefix(&mapsfolder)?
		} else {
			entry.strip_prefix(mapsfolder.parent().unwrap())?
		};
		let parent = parent.to_str().unwrap().replace('\\', "/");

		if mode == Mode::Automatic {
			let normalized = normalize_mapname(&stem);
			if let Some(existing_names_id) = existing_names.get(&normalized) {
				let in_recents = existing_recents.get(&normalized);
				if in_recents.is_none() || *in_recents.unwrap() != *existing_names_id {
					stem.insert(0, '#');
				}
			}
		} else if canon_clobber_check {
			if let Some(found) = existing_canon.get(&stem) {
				if digest != *found {
					// TODO: (low) canonClobber things...
					// canonClobber.write(f"{steam},{found},\n")
					// print_and_to_shit(f"  ^^^^ name collision {digest} {entry} (existing: {row[0]} & {found}")
				}
			}
		}

		outcsv.write_record([&stem, &filesize.to_string(), &filesize_bz2.to_string(), &digest, &parent])?;

		if !exists {
			newly_hashed.push(UnprocessedCsvRow {
				mapname: stem,
				filesize,
				filesize_bz2,
				sha1: digest,
				note: Some(parent),
				recently_added_note: None,
				datetime: None,
			});
		}
	}

	let mut f = tokio::fs::OpenOptions::new().append(true).open(&outcsvpath).await?;
	f.write_all(&outcsv.into_inner()?).await?;
	f.flush().await?;

	Ok(newly_hashed)
}
